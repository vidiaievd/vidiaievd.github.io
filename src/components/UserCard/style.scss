.user-card {
    border: 1px grey solid;
    border-radius: 10px;
    margin: 1rem auto;
    padding: 0.2rem 1rem;
    background-color: #e8e8e8;
    box-shadow: 5px 5px 5px rgba(0, 0, 255, 0.2);
    transform: scale(1);
    transition: all 0.3s ease-out;

    &.active {
        background-color: #b8b8ff;
    }

    &:hover {
        transform: scale(1.05);
    }

    h4 {
        margin: 0;
        color: black;

        &.active {
            color: blue;
        }
    }
}
