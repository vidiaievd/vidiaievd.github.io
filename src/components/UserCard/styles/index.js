export {StyledUserCard} from './StyledUserCard';
export {StyledTitleCard} from './StyledTitleCard';
export {StyledLabel} from './StyledLabel';