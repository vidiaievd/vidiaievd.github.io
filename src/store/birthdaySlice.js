import { createEntityAdapter, createSlice } from "@reduxjs/toolkit";

const birthdayAdapter = createEntityAdapter({
    sortComparer: (a, b) => (a.lastName > b.lastName ? 1 : -1),
});

export const {
    selectAll: selectBirthdays,
    selectById: selectBirthdaysById,
    selectIds,
} = birthdayAdapter.getSelectors((state) => state.birthday);

const stateFromStorage = () => {
    let entities = {};
    const ids = [];
    for (let i = 0; i < sessionStorage.length; i++) {
        let key = sessionStorage.key(i);
        entities[key] = JSON.parse(sessionStorage.getItem(key));
        ids.push(key);
    }
    return { entities, ids };
};

const birthdaySlice = createSlice({
    name: "birthday",
    initialState: birthdayAdapter.getInitialState({
        ids: stateFromStorage().ids,
        entities: stateFromStorage().entities,
    }),
    reducers: {
        addBirthday: {
            reducer: (state, action) => {
                birthdayAdapter.addOne(state, action.payload);
                sessionStorage.setItem(
                    action.payload.id,
                    JSON.stringify(action.payload)
                );
            },
        },
        removeBirthday: {
            reducer: (state, action) => {
                birthdayAdapter.removeOne(state, action.payload);
                sessionStorage.removeItem(action.payload);
            },
        },
        
    },
    extraReducers: {},
});

export const { addBirthday, removeBirthday, addToStorage } = birthdaySlice.actions;

export default birthdaySlice.reducer;
