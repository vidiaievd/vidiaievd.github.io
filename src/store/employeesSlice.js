import { createEntityAdapter, createSlice, createAsyncThunk} from "@reduxjs/toolkit";
import axios from "axios";

export const getEmployees = createAsyncThunk(
    "employees/getEmployees",
    async () => {
        const response = await axios.get(
            "https://yalantis-react-school-api.yalantis.com/api/task0/users"
        );
        if (response.status === 200) {
            return response.data;
        }
        return new Error("Error");
    }
);

const employeesAdapter = createEntityAdapter({
    sortComparer: (a, b) => (a.firstName > b.firstName ? 1 : -1),
});

export const { selectAll: selectEmployees, selectById: selectEmployeeById } =
    employeesAdapter.getSelectors((state) => state.employees);

const getEmployeesSlice = createSlice({
    name: "employees",
    initialState: employeesAdapter.getInitialState({
        loading: false,
        error: null,
    }),
    reducers: {},
    extraReducers: {
        [getEmployees.fulfilled](state, {payload}) {
            employeesAdapter.setAll(state, payload);
            state.loading = false;
        },
        [getEmployees.pending](state) {
            state.loading = true;
        },
        [getEmployees.rejected](state, action) {
            state.error = action.error.message;
            state.loading = false;
        },
    },
});

export default getEmployeesSlice.reducer;
