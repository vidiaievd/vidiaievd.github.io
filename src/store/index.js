import { combineReducers } from '@reduxjs/toolkit';
import employees from "../usersList/store/EmployeesSlice";
import birthday from "../usersList/store/BirthdaySlice";

const reducer = combineReducers({
  employees,
  birthday,
});

export default reducer;
