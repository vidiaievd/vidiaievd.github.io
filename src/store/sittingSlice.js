import { createSlice } from "@reduxjs/toolkit";

const sittingSlice = createSlice({
    name: "sitting",
    initialState: {
        language: JSON.parse(localStorage.getItem('language')) || 'en-US',
        theme: JSON.parse(localStorage.getItem('theme')) || 'light',
    },
    reducers: {
        language: {
            reducer: (state, { payload }) => {
                state.language = payload;
                localStorage.setItem('language', JSON.stringify(payload));
            },
        },
        theme: {
            reducer: (state, { payload }) => {
                state.theme = payload;
                localStorage.setItem('theme', JSON.stringify(payload));
            },
        },
    },
    extraReducers: {},
});

export const reducers = sittingSlice.actions;

export default sittingSlice.reducer;
