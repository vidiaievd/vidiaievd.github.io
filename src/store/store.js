import { configureStore } from '@reduxjs/toolkit';
import employeesReducer from './employeesSlice'
import birthdayReducer from './birthdaySlice';
import sittingReducer from './sittingSlice';

export default configureStore({
  reducer: {
    employees: employeesReducer,
    birthday: birthdayReducer,
    sitting: sittingReducer,
  },
});
