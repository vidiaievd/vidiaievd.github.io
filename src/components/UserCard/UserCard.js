import PT from "prop-types";
import { useDispatch, useSelector } from "react-redux";

import "./style.scss";

import { addBirthday, removeBirthday, selectIds } from "../../store/birthdaySlice";

export const UserCard = ({ employee }) => {
    const dispatch = useDispatch();
    const Ids = useSelector(selectIds);

    const handleSetDB = () => {
        dispatch(addBirthday(employee));
    };

    const handleDelDB = () => {
        dispatch(removeBirthday(employee.id));
    };

    const isChecked = (Ids.includes(employee.id));

    return (
        <div className={`user-card ${isChecked ? "active" : ""}`}>
            <h4 className={`user-card__name ${isChecked ? "active" : ""}`}>
                {`${employee.firstName} ${employee.lastName}`}
            </h4>
            <input
                id='not-active'
                type="radio"
                name={employee.id}
                value="false"
                defaultChecked={!isChecked}
                onChange={handleDelDB}
            />
            <label htmlFor="not-active">not active</label>
            <br />
            <input
                id='active'
                type="radio"
                name={employee.id}
                value="true"
                onChange={handleSetDB}
                defaultChecked={isChecked}
            />
            <label htmlFor="active">active</label>
        </div>
    );
};

UserCard.propTypes = {
    employee: PT.shape({
        id: PT.string.isRequired,
        firstName: PT.string.isRequired,
        lastName: PT.string.isRequired,
        dob: PT.string.isRequired,
    }).isRequired,
};
