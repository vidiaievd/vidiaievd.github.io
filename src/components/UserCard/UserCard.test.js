import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from "react-redux";
import { UserCard } from './UserCard';
import store from '../../store/store';

const employee = {
    id: "5e00928db89ff9c2559f9560",
    dob: "2019-05-09T03:24:32.532Z",
    firstName: "Watson",
    lastName: "Good",
}

describe('UserCard component', () => {
    it('UserCard renders', () => {
        render(
            <Provider store={store}>
                <UserCard employee={employee} />
            </Provider>);

        expect(screen.getByText(`${employee.firstName} ${employee.lastName}`)).toBeInTheDocument();
        expect(screen.getByText('active')).toBeInTheDocument();
        expect(screen.getByText('not active')).toBeInTheDocument();
        expect(screen.getByLabelText('not active')).toBeChecked();
        expect(screen.getByLabelText('active')).not.toBeChecked();
    });

    it('check active', () => {
        render(
            <Provider store={store}>
                <UserCard employee={employee} />
            </Provider>);

        expect(screen.getByLabelText('not active')).toBeChecked();
        expect(screen.getByLabelText('active')).not.toBeChecked();

        userEvent.click(screen.getByText('active'));
        expect(screen.getByLabelText('active')).toBeChecked();
        expect(screen.getByLabelText('not active')).not.toBeChecked();

        userEvent.click(screen.getByText('not active'));
        expect(screen.getByLabelText('not active')).toBeChecked();
        expect(screen.getByLabelText('active')).not.toBeChecked();
    });

    it('userCard snapshot', () => {
        const userCard = render(
            <Provider store={store}>
                <UserCard employee={employee} />
            </Provider>);

        expect(userCard).toMatchSnapshot();
    });

})