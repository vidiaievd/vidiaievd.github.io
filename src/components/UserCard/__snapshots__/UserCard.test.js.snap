// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard component userCard snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="user-card "
      >
        <h4
          class="user-card__name "
        >
          Watson Good
        </h4>
        <input
          checked=""
          id="not-active"
          name="5e00928db89ff9c2559f9560"
          type="radio"
          value="false"
        />
        <label
          for="not-active"
        >
          not active
        </label>
        <br />
        <input
          id="active"
          name="5e00928db89ff9c2559f9560"
          type="radio"
          value="true"
        />
        <label
          for="active"
        >
          active
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="user-card "
    >
      <h4
        class="user-card__name "
      >
        Watson Good
      </h4>
      <input
        checked=""
        id="not-active"
        name="5e00928db89ff9c2559f9560"
        type="radio"
        value="false"
      />
      <label
        for="not-active"
      >
        not active
      </label>
      <br />
      <input
        id="active"
        name="5e00928db89ff9c2559f9560"
        type="radio"
        value="true"
      />
      <label
        for="active"
      >
        active
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
